lib/shared/permissions.ts

```diff
@@ -0,0 +1,206 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import bindings = require('bindings');
+import * as Bluebird from 'bluebird';
+import * as childProcess from 'child_process';
+import { promises as fs } from 'fs';
+import * as _ from 'lodash';
+import * as os from 'os';
+import * as semver from 'semver';
+import * as sudoPrompt from 'sudo-prompt';
+import { promisify } from 'util';
+
+import { sudo as catalinaSudo } from './catalina-sudo/sudo';
+import * as errors from './errors';
+import { Dictionary, tmpFileDisposer } from './utils';
+
+const execAsync = promisify(childProcess.exec);
+const execFileAsync = promisify(childProcess.execFile);
+const sudoExecAsync = promisify(sudoPrompt.exec);
+
+/**
+ * @summary The user id of the UNIX "superuser"
+ */
+const UNIX_SUPERUSER_USER_ID = 0;
+
+export async function isElevated(): Promise<boolean> {
+	if (os.platform() === 'win32') {
+		// `fltmc` is available on WinPE, XP, Vista, 7, 8, and 10
+		// Works even when the "Server" service is disabled
+		// See http://stackoverflow.com/a/28268802
+		try {
+			await execAsync('fltmc');
+		} catch (error) {
+			if (error.code === os.constants.errno.EPERM) {
+				return false;
+			}
+			throw error;
+		}
+		return true;
+	}
+	return process.geteuid() === UNIX_SUPERUSER_USER_ID;
+}
+
+/**
+ * @summary Check if the current process is running with elevated permissions
+ */
+export function isElevatedUnixSync(): boolean {
+	return process.geteuid() === UNIX_SUPERUSER_USER_ID;
+}
+
+function escapeSh(value: any): string {
+	// Make sure it's a string
+	// Replace ' -> '\'' (closing quote, escaped quote, opening quote)
+	// Surround with quotes
+	return `'${String(value).replace(/'/g, "'\\''")}'`;
+}
+
+function escapeParamCmd(value: any): string {
+	// Make sure it's a string
+	// Escape " -> \"
+	// Surround with double quotes
+	return `"${String(value).replace(/"/g, '\\"')}"`;
+}
+
+function setEnvVarSh(value: any, name: string): string {
+	return `export ${name}=${escapeSh(value)}`;
+}
+
+function setEnvVarCmd(value: any, name: string): string {
+	return `set "${name}=${String(value)}"`;
+}
+
+// Exported for tests
+export function createLaunchScript(
+	command: string,
+	argv: string[],
+	environment: Dictionary<string>,
+): string {
+	const isWindows = os.platform() === 'win32';
+	const lines = [];
+	if (isWindows) {
+		// Switch to utf8
+		lines.push('chcp 65001');
+	}
+	const [setEnvVarFn, escapeFn] = isWindows
+		? [setEnvVarCmd, escapeParamCmd]
+		: [setEnvVarSh, escapeSh];
+	lines.push(..._.map(environment, setEnvVarFn));
+	lines.push([command, ...argv].map(escapeFn).join(' '));
+	return lines.join(os.EOL);
+}
+
+async function elevateScriptWindows(
+	path: string,
+): Promise<{ cancelled: boolean }> {
+	// 'elevator' imported here as it only exists on windows
+	// TODO: replace this with sudo-prompt once https://github.com/jorangreef/sudo-prompt/issues/96 is fixed
```

Comment:
https://github.com/jorangreef/sudo-prompt/issues/96 claims to be fixed now?

Reply:
@lurch, thanks, noted, that will be for another PR.